File: tradex/memory/policy.ts
export class MemoryRuntimePolicy {
  readonly enabled: boolean;
  readonly generateMemories: boolean;
  readonly useMemories: boolean;
  readonly ephemeral: boolean;
  readonly disableOnExternalContext: boolean;

  private constructor(input: { enabled: boolean; generateMemories: boolean; useMemories: boolean; ephemeral: boolean; disableOnExternalContext?: boolean }) {
    this.enabled = input.enabled;
    this.generateMemories = input.generateMemories;
    this.useMemories = input.useMemories;
    this.ephemeral = input.ephemeral;
    this.disableOnExternalContext = input.disableOnExternalContext ?? false;
  }

  static normal(): MemoryRuntimePolicy {
    return new MemoryRuntimePolicy({ enabled: true, generateMemories: true, useMemories: true, ephemeral: false });
  }

  static consolidation(): MemoryRuntimePolicy {
    return new MemoryRuntimePolicy({ enabled: true, generateMemories: false, useMemories: false, ephemeral: true });
  }

  static disabled(): MemoryRuntimePolicy {
    return new MemoryRuntimePolicy({ enabled: false, generateMemories: false, useMemories: false, ephemeral: false });
  }
}

File: tradex/memory/policy.py
"""文件用途：控制一次 agent/runtime 是否允许读写本地记忆。"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class MemoryRuntimePolicy:
    """说明：tradex 版的 Codex memory 运行策略。

    `ephemeral` 不是 OS 级沙箱；它只表达“这个临时 worker 不能读记忆、
    也不能生成新记忆”，用来切断 consolidation 之类内部任务的递归入口。
    """

    ephemeral: bool = False
    generate_memories: bool = True
    use_memories: bool = True

    def __post_init__(self) -> None:
        if self.ephemeral:
            object.__setattr__(self, "generate_memories", False)
            object.__setattr__(self, "use_memories", False)

    @classmethod
    def normal(cls) -> "MemoryRuntimePolicy":
        """说明：普通交互会话，允许读已有记忆，也允许后台生成记忆。"""
        return cls()

    @classmethod
    def consolidation(cls) -> "MemoryRuntimePolicy":
        """说明：Phase 2 内部整理 worker，禁止递归读写记忆。"""
        return cls(ephemeral=True)

    @classmethod
    def disabled(cls) -> "MemoryRuntimePolicy":
        """说明：显式关闭 memory 的读写入口。"""
        return cls(generate_memories=False, use_memories=False)


__all__ = ["MemoryRuntimePolicy"]
